### PR TITLE
Develop grid node

### PR DESCRIPTION
**Pull request by dabingnn**

### CODE REVIEW

File: cocos/2d/platform/android/java/project.properties [hunk 1]
```diff
@@ -12,4 +12,4 @@
 
 android.library=true
 # Project target.
-target=android-10
+target=android-17
```
Comment (darkdukey) [inline cocos/2d/platform/android/java/project.properties]: Why do we need to change android target here? android-17 means 4.2 +

Comment (minggo) [inline cocos/2d/platform/android/java/project.properties]: I think it is a mistake.
Should not change it.

Comment (winipcfg) [inline cocos/2d/platform/android/java/project.properties]: https://github.com/cocos2d/cocos2d-x/blob/develop/template/multi-platform-cpp/proj.android/AndroidManifest.xml
http://developer.android.com/guide/topics/manifest/activity-element.html

Isn't that the options for "android:configChanges" require at least API 13?

Comment (dabingnn) [inline cocos/2d/platform/android/java/project.properties]: The changes are made automatically by eclipse. I am sorry for committing it by accident.I have reversed the changes to these two files.

File: cocos/2d/CCActionGrid.cpp [hunk 2]
```diff
@@ -345,14 +352,20 @@ DeccelAmplitude* DeccelAmplitude::reverse() const
...
+    GridBase *grid = _gridNodeTarget->getGrid();
     if (grid && grid->isActive())
     {
         grid->setActive(false);
     }
 }
 
+void StopGrid::cacheTargetAsGridNode()
+{
+    _gridNodeTarget = dynamic_cast<NodeGrid*> (_target);
```
Comment (minggo) [inline cocos/2d/CCActionGrid.cpp:365]: Can we change the type of `_target` to `NodeGrid`, then don't have to dynamic cast.

Comment (minggo) [inline cocos/2d/CCActionGrid.cpp:365]: Ok, it is in action. Ignore me.

File: cocos/2d/CCNode.cpp [hunk 3]
```diff
@@ -872,13 +851,12 @@ void Node::transform()
...
 
     kmGLMultMatrix( &transfrom4x4 );
-
     // saves the MV matrix
     kmGLGetMatrix(KM_GL_MODELVIEW, &_modelViewTransform);
 
-
     // XXX: Expensive calls. Camera should be integrated into the cached affine matrix
-    if ( _camera != nullptr && !(_grid != nullptr && _grid->isActive()) )
+    //_grid is always null
```
Comment (minggo) [inline cocos/2d/CCNode.cpp]: I think this comment should be removed. Because now Node don't have _grid.

File: cocos/2d/CCNodeGrid.cpp [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+#include "CCRenderer.h"
+#include "CCCustomCommand.h"
+
+
+NS_CC_BEGIN
+
+NodeGrid* NodeGrid::create()
+{
+    NodeGrid * pRet = new NodeGrid();
+    if (pRet && pRet->init())
```
Comment (minggo) [inline cocos/2d/CCNodeGrid.cpp]: Don't use hungarian style.

File: cocos/2d/CCNodeGrid.cpp [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    }
+    else
+    {
+        CC_SAFE_DELETE(pRet);
+    }
+    return pRet;
+}
+
+NodeGrid::NodeGrid()
+:Node()
```
Comment (minggo) [inline cocos/2d/CCNodeGrid.cpp]: Why should invoke `Node()` manually?

File: cocos/2d/CCNodeGrid.cpp [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+    Director::Projection beforeProjectionType;
+    if(_nodeGrid && _nodeGrid->isActive())
+    {
+        beforeProjectionType = Director::getInstance()->getProjection();
+        _nodeGrid->set2DProjection();
+    }
+    
+    kmGLGetMatrix(KM_GL_MODELVIEW, &_cachedMVmat);
+    
+    CustomCommand* gridBeginCmd = CustomCommand::getCommandPool().generateCommand();
```
Comment (minggo) [inline cocos/2d/CCNodeGrid.cpp:93]: `CustomCommand::getCommandPool()` should be return a pointer to obey our coding style.

File: cocos/2d/CCTransition.cpp [hunk 7]
```diff
@@ -1346,10 +1347,12 @@ void TransitionCrossFade::onExit()
 //
 TransitionTurnOffTiles::TransitionTurnOffTiles()
 {
+    _outSceneProxy = nullptr;
```
Comment (minggo) [inline cocos/2d/CCTransition.cpp]: Why not use initialization list?

File: cocos/2d/CCTransition.cpp [hunk 8]
```diff
@@ -1516,6 +1572,10 @@ void TransitionFadeTR::sceneOrder()
 void TransitionFadeTR::onEnter()
 {
     TransitionScene::onEnter();
+    _outSceneProxy = NodeGrid::create();
+    _outSceneProxy->retain();
+    _outSceneProxy->setTarget(_outScene);
+    _outSceneProxy->onEnter();
```
Comment (minggo) [inline cocos/2d/CCTransition.cpp:1577]: Why create `NodeGrid` in `onEnter()`, and released in destructor?

File: cocos/2d/CCNodeGrid.cpp (hunk 6 above)
Comment (dabingnn) [inline cocos/2d/CCNodeGrid.cpp:93]: The command pool will be removed in single  thread version. So I will postpone the changes and remove them directly.

Comment (minggo) [inline cocos/2d/CCNodeGrid.cpp:93]: Ok.

File: cocos/2d/CCActionGrid.cpp [hunk 9]
```diff
@@ -97,19 +104,19 @@ GridBase* Grid3DAction::getGrid()
 
 Vertex3F Grid3DAction::getVertex(const Point& position) const
 {
-    Grid3D *g = (Grid3D*)_target->getGrid();
+    Grid3D *g = (Grid3D*)_gridNodeTarget->getGrid();
```
Comment (ricardoquesada) [inline cocos/2d/CCActionGrid.cpp:107]: and `static_cast<>` is better, but we can change it later

Comment (dabingnn) [inline cocos/2d/CCActionGrid.cpp:107]: Thanks ,I will fix this in another PR

**Pull request merged**